pkg/source/ttnv3/source.go

```diff
@@ -0,0 +1,321 @@
+package ttnv3
+
+import (
+	"context"
+
+	"github.com/spf13/pflag"
+	"go.thethings.network/lorawan-stack/v3/pkg/ttnpb"
+
+	"go.thethings.network/lorawan-stack-migrate/pkg/source"
+	"go.thethings.network/lorawan-stack-migrate/pkg/source/ttnv3/api"
+)
+
+// Source implements the Source interface.
+type Source struct {
+	ctx context.Context
+
+	config *config
+}
+
+// NewSource creates a new TTNv3 cource
+func NewSource(ctx context.Context, flags *pflag.FlagSet) (source.Source, error) {
+	config, err := getConfig(flags)
+	if err != nil {
+		return Source{}, err
+	}
+	s := Source{
+		ctx:    ctx,
+		config: config,
+	}
+	return s, nil
+}
+
+// ExportDevice implements the source.Source interface.
+func (s Source) ExportDevice(devID string) (*ttnpb.EndDevice, error) {
+	isPaths, nsPaths, asPaths, jsPaths := splitEndDeviceGetPaths(ttnpb.BottomLevelFields(ttnpb.EndDeviceFieldPathsNested)...)
+	if len(nsPaths) > 0 {
+		isPaths = ttnpb.AddFields(isPaths, "network_server_address")
+	}
+	if len(asPaths) > 0 {
+		isPaths = ttnpb.AddFields(isPaths, "application_server_address")
+	}
+	if len(jsPaths) > 0 {
+		isPaths = ttnpb.AddFields(isPaths, "join_server_address")
+	}
+	is, err := api.Dial(s.ctx, s.config.identityServerGRPCAddress)
+	if err != nil {
+		return nil, err
+	}
+	ids := &ttnpb.EndDeviceIdentifiers{
+		ApplicationIds: &ttnpb.ApplicationIdentifiers{ApplicationId: s.config.appID},
+		DeviceId:       devID,
+	}
+	dev, err := ttnpb.NewEndDeviceRegistryClient(is).Get(s.ctx, &ttnpb.GetEndDeviceRequest{
+		EndDeviceIds: ids,
+		FieldMask:    ttnpb.FieldMask(isPaths...),
+	})
+	if err != nil {
+		return nil, err
+	}
+	if dev.ClaimAuthenticationCode.GetValue() != "" {
+		// ClaimAuthenticationCode is already retrieved from the IS. We can unset the related JS paths
+		jsPaths = ttnpb.ExcludeFields(jsPaths, claimAuthenticationCodePaths...)
+	}
+	res, err := s.getEndDevice(ids, nsPaths, asPaths, jsPaths, true)
+	if err != nil {
+		return nil, err
+	}
+	paths := ttnpb.AddFields(nsPaths, ttnpb.AddFields(asPaths, ttnpb.AddFields(jsPaths, "ids.dev_addr")...)...)
+	if err := dev.SetFields(res, paths...); err != nil {
+		return nil, err
+	}
+	// Clear all "{server}_address" fields from exported device
+	if err := dev.SetFields(&ttnpb.EndDevice{}, "application_server_address", "join_server_address", "network_server_address"); err != nil {
```

Comment:
I think even using `nil` as first parameter is supported.

Reply:
Works indeed, fixed